src/util/string2int.cpp

```diff
@@ -89,3 +91,77 @@ unsigned long long int unsafe_string2unsignedlonglong(
 {
   return str2number<unsigned long long int>(str.c_str(), base, false);
 }
+
+template <typename do_conversiont>
+auto wrap_string_conversion(do_conversiont do_conversion)
+  -> optionalt<decltype(do_conversion())>
+{
+  try
+  {
+    return do_conversion();
+  }
+  catch(const std::invalid_argument &)
+  {
+    return nullopt;
+  }
+  catch(const std::out_of_range &)
+  {
+    return nullopt;
+  }
+}
+
+optionalt<int> string2optional_int(const std::string &str, int base)
+{
+  return wrap_string_conversion(
+    [&]() { return std::stoi(str, nullptr, base); });
+}
+
+optionalt<unsigned> string2optional_unsigned(const std::string &str, int base)
+{
+  return wrap_string_conversion([&]() {
+    if(str.find('-') != std::string::npos)
+    {
+      // stoul wraps around if given negative numbers,
+      // but since we need to narrow cast this leads to weird behavior
+      // so we disallow it here
+      throw std::out_of_range{"negative values not allowed"};
+    }
+    auto const ul_value = std::stoul(str, nullptr, base);
+    auto const u_value = narrow_cast<unsigned>(ul_value);
+    if(u_value == ul_value)
+    {
+      return u_value;
+    }
+    else
+    {
+      throw std::out_of_range{"conversion out of range"};
```

Comment:
Why throw and then wrap the function in a try catch to return an optional? Return an optional directly.

Reply:
Because `sto*` type functions throw on failure. The function that's passed into `wrap` is basically just `sto`, but I need to do some additional checks on the string before passing it to `stol` as well as checking the conversion result and it makes sense to just use the same interface for that.

Reply:
I would prefer wrapping just the std::stoul in that case. The string in the exception here is never used and so is just a comment. That's misleading because we would expect it to be useful somewhere.